File: solutions/2026-07-11/java/2026-07-11-count-the-number-of-complete-components.java
```java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

// Problem: Count the Number of Complete Components
// Link: https://leetcode.com/problems/count-the-number-of-complete-components/
// Approach:
// We can use Depth First Search (DFS) or Breadth First Search (BFS) to find connected components.
// For each connected component found, we need to verify if it's a "complete" component.
// A component is complete if the number of edges within that component is equal to k * (k - 1) / 2,
// where k is the number of vertices in that component.
// We can maintain an adjacency list to represent the graph. During DFS/BFS, we keep track of visited nodes,
// the number of nodes in the current component, and the number of edges encountered within the component.
//
// Steps:
// 1. Build the adjacency list representation of the graph.
// 2. Initialize a boolean array `visited` to keep track of visited vertices.
// 3. Initialize `completeComponentsCount` to 0.
// 4. Iterate through each vertex from 0 to n-1.
// 5. If a vertex is not visited, start a DFS/BFS from it to find a connected component.
// 6. During DFS/BFS:
//    a. Count the number of vertices (`numNodes`) in the current component.
//    b. Count the number of edges (`numEdges`) in the current component. Be careful not to double count edges.
//       A simple way is to sum the degrees of all nodes in the component and divide by 2.
//    c. Mark all visited nodes within the component.
// 7. After exploring a component, check if `numEdges == numNodes * (numNodes - 1) / 2`.
// 8. If the condition in step 7 is true, increment `completeComponentsCount`.
// 9. Return `completeComponentsCount`.
//
// Time Complexity:
// Building the adjacency list takes O(E) time, where E is the number of edges.
// The DFS/BFS traversal visits each vertex and edge at most once. So, O(V + E), where V is the number of vertices.
// The check for completeness takes O(1) per component.
// Therefore, the overall time complexity is O(V + E). Given n <= 50, this is very efficient.
//
// Space Complexity:
// The adjacency list takes O(V + E) space.
// The `visited` array takes O(V) space.
// The recursion stack for DFS (or queue for BFS) can take up to O(V) space in the worst case.
// Therefore, the overall space complexity is O(V + E).
class Solution {
    // Adjacency list to store the graph
    private List<List<Integer>> adj;
    // Array to keep track of visited nodes during DFS
    private boolean[] visited;
    // Variables to store the count of nodes and edges in the current component
    private int currentComponentNodes;
    private int currentComponentEdges;

    /**
     * Counts the number of complete connected components in a graph.
     * A connected component is complete if there is an edge between every pair of its vertices.
     *
     * @param n     The number of vertices in the graph.
     * @param edges A 2D array representing the edges of the graph.
     * @return The number of complete connected components.
     */
    public int countCompleteComponents(int n, int[][] edges) {
        // Initialize adjacency list and visited array
        adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        visited = new boolean[n];

        // Build the adjacency list from the given edges
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }

        int completeComponentsCount = 0;

        // Iterate through all vertices to find connected components
        for (int i = 0; i < n; i++) {
            // If a vertex has not been visited, it means it's part of a new component
            if (!visited[i]) {
                // Reset counts for the new component
                currentComponentNodes = 0;
                currentComponentEdges = 0;

                // Start DFS to explore the connected component
                dfs(i);

                // After exploring the component, check if it's complete
                // A component with k nodes is complete if it has k * (k - 1) / 2 edges
                if (currentComponentEdges == currentComponentNodes * (currentComponentNodes - 1) / 2) {
                    completeComponentsCount++;
                }
            }
        }

        return completeComponentsCount;
    }

    /**
     * Performs Depth First Search (DFS) to find a connected component and count its nodes and edges.
     *
     * @param u The current vertex being visited.
     */
    private void dfs(int u) {
        // Mark the current node as visited
        visited[u] = true;
        // Increment the count of nodes in the current component
        currentComponentNodes++;
        // Add the degree of the current node to the edge count (will be divided by 2 later)
        currentComponentEdges += adj.get(u).size();

        // Iterate through all neighbors of the current node
        for (int v : adj.get(u)) {
            // If a neighbor has not been visited, recursively call DFS on it
            if (!visited[v]) {
                dfs(v);
            }
        }
    }
}
```

File: solutions/2026-07-11/cpp/2026-07-11-count-the-number-of-complete-components.cpp
```cpp
// Problem: Count the Number of Complete Components
// Link: https://leetcode.com/problems/count-the-number-of-complete-components/
// Approach:
// We can solve this problem by iterating through each connected component of the graph.
// For each component, we need to check if it's a "complete" component.
// A connected component is complete if there's an edge between every pair of vertices within that component.
//
// We can use Depth-First Search (DFS) or Breadth-First Search (BFS) to find connected components.
// During the traversal of a component, we can:
// 1. Count the number of vertices in the component.
// 2. Count the number of edges within the component.
// 3. Keep track of all vertices belonging to the current component.
//
// After identifying a component, we check if it's complete.
// A component with 'V' vertices is complete if it has exactly V * (V - 1) / 2 edges.
// We also need to ensure that all pairs of vertices within the component are indeed connected by an edge.
//
// An alternative and more efficient approach involves using Union-Find to group vertices into components.
// After building the Union-Find structure, we can iterate through each component:
// 1. For each component representative, find all vertices belonging to it.
// 2. Count the number of vertices (V) in this component.
// 3. Count the number of edges (E) where both endpoints belong to this component.
// 4. Check if E == V * (V - 1) / 2. If it is, increment the complete component count.
//
// To efficiently count edges within a component using Union-Find:
// We can iterate through all given edges. For each edge [u, v], if find(u) == find(v),
// it means u and v belong to the same component. We then increment the edge count for that component.
// We can use a map or an array to store edge counts per component representative.
//
// Let's detail the Union-Find approach:
// 1. Initialize a Union-Find data structure for 'n' vertices.
// 2. Iterate through the `edges` array. For each edge `[u, v]`, call `union_sets(u, v)`.
//    This groups connected vertices into the same set.
// 3. Create two maps:
//    - `component_vertex_count`: Stores the number of vertices for each component representative.
//    - `component_edge_count`: Stores the number of edges within each component representative.
// 4. Iterate through the `edges` array again. For each edge `[u, v]`:
//    - Find the representative of `u` (which will be the same as `v` since they are connected): `root = find(u)`.
//    - Increment `component_edge_count[root]`.
// 5. Iterate through all vertices from 0 to n-1. For each vertex `i`:
//    - Find its representative: `root = find(i)`.
//    - Increment `component_vertex_count[root]`. This ensures we count all vertices, including isolated ones.
// 6. Initialize `complete_components_count = 0`.
// 7. Iterate through the `component_vertex_count` map (or iterate through representatives if you tracked them).
//    For each representative `root` and its vertex count `V`:
//    - Get the edge count `E` from `component_edge_count[root]` (if `root` is not in the map, `E` is 0).
//    - Calculate the expected number of edges for a complete component of size `V`: `expected_edges = V * (V - 1) / 2`.
//    - If `E == expected_edges`, then this component is complete. Increment `complete_components_count`.
//
// Time Complexity:
// - Union-Find initialization: O(n)
// - Processing edges for union: O(E * alpha(n)), where E is the number of edges and alpha is the inverse Ackermann function (very slow-growing, practically constant).
// - Counting edges per component: O(E * alpha(n))
// - Counting vertices per component: O(n * alpha(n))
// - Checking completeness: O(n) in the worst case (if every vertex is a component representative).
// Overall: O(n + E * alpha(n)), which is effectively O(n + E). Since E can be up to n*(n-1)/2, it's O(n^2).
// Given n <= 50, O(n^2) is acceptable.
//
// Space Complexity:
// - Parent array for Union-Find: O(n)
// - Rank/Size array for Union-Find: O(n)
// - `component_vertex_count` map: O(n) in the worst case (all distinct components)
// - `component_edge_count` map: O(n) in the worst case
// Overall: O(n).
//
// Alternative DFS/BFS approach:
// 1. Build an adjacency list for the graph: O(n + E).
// 2. Initialize a `visited` array.
// 3. Iterate through all vertices from 0 to n-1.
// 4. If a vertex `i` is not visited:
//    a. Start a DFS/BFS from `i` to find a connected component.
//    b. During the traversal, count the number of vertices (`V`) and edges (`E_component`) within this component.
//       To count edges correctly without double-counting, we can iterate through the adjacency list of each visited node within the component.
//       A simpler way is to count the total degree sum of nodes in the component and divide by 2.
//    c. Mark all visited nodes.
//    d. Check if `E_component == V * (V - 1) / 2`. If so, increment the complete component count.
//
// Time Complexity (DFS/BFS):
// - Building adjacency list: O(n + E)
// - DFS/BFS traversals: Each vertex and edge is visited at most once across all traversals. So, O(n + E).
// - Checking completeness within each component: For a component of size V, we might iterate through its vertices and their adjacency lists. This can be done efficiently within the DFS/BFS.
//   The total work for checking completeness is bounded by the total degree sum, which is 2*E.
// Overall: O(n + E).
//
// Space Complexity (DFS/BFS):
// - Adjacency list: O(n + E)
// - Visited array: O(n)
// - Recursion stack (for DFS) or queue (for BFS): O(n)
// Overall: O(n + E).
//
// Given the constraints (n <= 50), both Union-Find and DFS/BFS are viable.
// Union-Find might be slightly more conceptually clean for counting component properties.
// Let's proceed with the Union-Find approach for its elegance in grouping and then checking properties.

#include <vector>
#include <numeric>
#include <unordered_map>
#include <unordered_set>

// Union-Find structure
struct UnionFind {
    std::vector<int> parent;
    std::vector<int> size; // Using size to optimize union by size

    // Constructor: Initializes n disjoint sets
    UnionFind(int n) {
        parent.resize(n);
        std::iota(parent.begin(), parent.end(), 0); // Fill with 0, 1, 2, ... n-1
        size.assign(n, 1); // Each set initially has size 1
    }

    // Find operation with path compression
    int find(int i) {
        if (parent[i] == i) {
            return i;
        }
        return parent[i] = find(parent[i]); // Path compression
    }

    // Union operation by size
    void union_sets(int i, int j) {
        int root_i = find(i);
        int root_j = find(j);

        if (root_i != root_j) {
            // Union by size: attach smaller tree to larger tree
            if (size[root_i] < size[root_j]) {
                std::swap(root_i, root_j);
            }
            parent[root_j] = root_i;
            size[root_i] += size[root_j];
        }
    }
};

class Solution {
public:
    int countCompleteComponents(int n, std::vector<std::vector<int>>& edges) {
        // Initialize Union-Find structure
        UnionFind uf(n);

        // Process all edges to form connected components
        for (const auto& edge : edges) {
            uf.union_sets(edge[0], edge[1]);
        }

        // Maps to store counts for each component representative
        // `component_vertex_count[root]` -> number of vertices in the component rooted at `root`
        // `component_edge_count[root]` -> number of edges within the component rooted at `root`
        std::unordered_map<int, int> component_vertex_count;
        std::unordered_map<int, int> component_edge_count;

        // Iterate through all vertices to populate vertex counts for each component
        // This also correctly handles isolated vertices as components of size 1.
        for (int i = 0; i < n; ++i) {
            int root = uf.find(i);
            component_vertex_count[root]++;
        }

        // Iterate through all edges to populate edge counts for each component
        for (const auto& edge : edges) {
            int u = edge[0];
            // Since u and v are connected, they will have the same root.
            int root = uf.find(u);
            component_edge_count[root]++;
        }

        int complete_components_count = 0;

        // Iterate through the unique component representatives (keys in component_vertex_count)
        for (auto const& [root, num_vertices] : component_vertex_count) {
            // Get the number of edges for this component. If the root has no edges, this map will not contain it,
            // so `count` will be 0 by default if `find` returns an iterator to `end()`.
            int num_edges = 0;
            if (component_edge_count.count(root)) {
                num_edges = component_edge_count[root];
            }

            // Calculate the expected number of edges for a complete component of size `num_vertices`.
            // A complete graph with V vertices has V*(V-1)/2 edges.
            long long expected_edges = (long long)num_vertices * (num_vertices - 1) / 2;

            // Check if the component is complete:
            // 1. The number of edges must match the expected number for a complete graph.
            // 2. Note: The Union-Find structure ensures connectivity. If the edge count matches the
            //    expected count for a complete graph, it implies all pairs are connected.
            if (num_edges == expected_edges) {
                complete_components_count++;
            }
        }

        return complete_components_count;
    }
};
```